// radiks/lib/models/group-invitation.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _model = _interopRequireDefault(require("../model"));

var _user = _interopRequireDefault(require("./user"));

var _groupMembership = _interopRequireDefault(require("./group-membership"));

var _helpers = require("../helpers");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _possibleConstructorReturn(self, call) { if (call && (typeof call === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

let GroupInvitation =
/*#__PURE__*/
function (_Model) {
  _inherits(GroupInvitation, _Model);

  function GroupInvitation(...args) {
    var _this;

    _classCallCheck(this, GroupInvitation);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(GroupInvitation).call(this, ...args));

    _defineProperty(_assertThisInitialized(_assertThisInitialized(_this)), "userPublicKey", void 0);

    return _this;
  }

  _createClass(GroupInvitation, [{
    key: "activate",
    value: async function activate() {
      const {
        userGroups
      } = (0, _helpers.userGroupKeys)();
      const groupId = this.attrs.userGroupId;

      if (userGroups[groupId]) {
        return true;
      }

      const groupMembership = new _groupMembership.default({
        userGroupId: this.attrs.userGroupId,
        username: (0, _helpers.loadUserData)().username,
        signingKeyPrivateKey: this.attrs.signingKeyPrivateKey,
        signingKeyId: this.attrs.signingKeyId
      });
      await groupMembership.save();
      await _groupMembership.default.cacheKeys();
      return groupMembership;
    }
  }, {
    key: "encryptionPublicKey",
    value: async function encryptionPublicKey() {
      return this.userPublicKey;
    }
  }, {
    key: "encryptionPrivateKey",
    value: function encryptionPrivateKey() {
      return (0, _helpers.loadUserData)().appPrivateKey;
    }
  }], [{
    key: "makeInvitation",
    value: async function makeInvitation(username, userGroup) {
      const user = new _user.default({
        _id: username
      });
      await user.fetch({
        decrypt: false
      });
      const {
        publicKey
      } = user.attrs;
      const invitation = new this({
        userGroupId: userGroup._id,
        signingKeyPrivateKey: userGroup.privateKey,
        signingKeyId: userGroup.attrs.signingKeyId
      });
      invitation.userPublicKey = publicKey;
      await invitation.save();
      return invitation;
    }
  }]);

  return GroupInvitation;
}(_model.default);

exports.default = GroupInvitation;

_defineProperty(GroupInvitation, "className", 'GroupInvitation');

_defineProperty(GroupInvitation, "schema", {
  userGroupId: String,
  signingKeyPrivateKey: String,
  signingKeyId: String
});

_defineProperty(GroupInvitation, "defaults", {
  updatable: false
});
