// radiks/lib/helpers.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.loadUserData = exports.requireUserSession = exports.addUserGroupKey = exports.addPersonalSigningKey = exports.userGroupKeys = exports.clearStorage = exports.encryptObject = exports.decryptObject = exports.GROUP_MEMBERSHIPS_STORAGE_KEY = void 0;

var _encryption = require("blockstack/lib/encryption");

var _config = require("./config");

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; var ownKeys = Object.keys(source); if (typeof Object.getOwnPropertySymbols === 'function') { ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function (sym) { return Object.getOwnPropertyDescriptor(source, sym).enumerable; })); } ownKeys.forEach(function (key) { _defineProperty(target, key, source[key]); }); } return target; }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

const GROUP_MEMBERSHIPS_STORAGE_KEY = 'GROUP_MEMBERSHIPS_STORAGE_KEY';
exports.GROUP_MEMBERSHIPS_STORAGE_KEY = GROUP_MEMBERSHIPS_STORAGE_KEY;

const valueToString = (value, clazz) => {
  if (clazz === Boolean) {
    return value ? 'true' : 'false';
  }

  if (clazz === Number) {
    return String(value);
  }

  if (clazz === Array || clazz === Object) {
    return JSON.stringify(value);
  }

  return value;
};

const stringToValue = (value, clazz) => {
  if (clazz === Boolean) {
    return value === 'true';
  }

  if (clazz === Number) {
    return parseFloat(value);
  }

  if (clazz === Array || clazz === Object) {
    return JSON.parse(value);
  }

  return value;
};

const decryptObject = async (encrypted, model) => {
  const privateKey = await model.encryptionPrivateKey();

  const decrypted = _objectSpread({}, encrypted);

  const {
    schema
  } = model;
  Object.keys(encrypted).forEach(key => {
    const value = encrypted[key];
    const schemaValue = schema[key];
    let clazz = schemaValue;
    const schemaAttribute = schema[key];

    if (schemaAttribute && schemaAttribute.type) {
      clazz = schemaAttribute.type;
    }

    if (clazz && schemaAttribute && !schemaAttribute.decrypted) {
      try {
        const decryptedValue = (0, _encryption.decryptECIES)(privateKey, value);
        decrypted[key] = stringToValue(decryptedValue, clazz);
      } catch (error) {
        console.debug(`Decryption error for key: '${key}': ${error.message}`); // eslint-disable-line

        decrypted[key] = value;
      }
    }
  });
  return decrypted;
};

exports.decryptObject = decryptObject;

const encryptObject = async model => {
  const publicKey = await model.encryptionPublicKey();
  const object = model.attrs;

  const encrypted = _objectSpread({}, object, {
    _id: model._id
  });

  Object.keys(model.schema).forEach(key => {
    const schemaValue = model.schema[key];
    const schemaAttribute = model.schema[key];
    const value = object[key];
    let clazz = schemaValue;
    if (typeof value === 'undefined') return;

    if (schemaAttribute.type) {
      clazz = schemaAttribute.type;
    }

    if (schemaAttribute.decrypted) {
      encrypted[key] = value;
      return;
    }

    const stringValue = valueToString(value, clazz);
    encrypted[key] = (0, _encryption.encryptECIES)(publicKey, stringValue);
  });
  return encrypted;
};

exports.encryptObject = encryptObject;

const clearStorage = () => {
  localStorage.removeItem(GROUP_MEMBERSHIPS_STORAGE_KEY);
};

exports.clearStorage = clearStorage;

const userGroupKeys = () => {
  const keysString = localStorage.getItem(GROUP_MEMBERSHIPS_STORAGE_KEY);
  let keys = keysString ? JSON.parse(keysString) : {};
  keys = _objectSpread({
    userGroups: {},
    signingKeys: {},
    personal: {}
  }, keys);
  return keys;
};

exports.userGroupKeys = userGroupKeys;

const addPersonalSigningKey = signingKey => {
  const keys = userGroupKeys();
  keys.personal = _objectSpread({
    _id: signingKey._id
  }, signingKey.attrs);
  localStorage.setItem(GROUP_MEMBERSHIPS_STORAGE_KEY, JSON.stringify(keys));
};

exports.addPersonalSigningKey = addPersonalSigningKey;

const addUserGroupKey = userGroup => {
  const keys = userGroupKeys();
  keys.userGroups[userGroup._id] = userGroup.attrs.signingKeyId;
  keys.signingKeys[userGroup.attrs.signingKeyId] = userGroup.privateKey;
  localStorage.setItem(GROUP_MEMBERSHIPS_STORAGE_KEY, JSON.stringify(keys));
};

exports.addUserGroupKey = addUserGroupKey;

const requireUserSession = () => {
  const {
    userSession
  } = (0, _config.getConfig)();

  if (!userSession) {
    // TODO: link to docs
    throw new Error('You have not properly configured your UserSession.');
  }

  return userSession;
};

exports.requireUserSession = requireUserSession;

const loadUserData = () => {
  const {
    userSession
  } = (0, _config.getConfig)();

  if (userSession) {
    return userSession.loadUserData();
  }

  return null;
};

exports.loadUserData = loadUserData;
